src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs

```diff
@@ -578,6 +565,112 @@ internal bool IsUnsuppressableError()
         /// </summary>
         internal bool IsUnsuppressedError
             => Severity == DiagnosticSeverity.Error && !IsSuppressed;
+
+        /// <summary>
+        /// Implements the following enumerator without allocations:
+        ///
+        /// <code>
+        /// if (this.Location.SourceTree == tree)
+        /// {
+        ///     yield return this.Location;
+        /// }
+        /// 
+        /// if (this.AdditionalLocations != null)
```

Comment:
📝 It is rare for AdditionalLocations to be non-empty. An alternative to this pull request is to just add a fast-path to `HasIntersectingLocation` for the case where it is empty.